### PR TITLE
roachtest: improve artifacts collection on timeout

### PR DESCRIPTION
**Pull request by tbg**
Previously, when a roachtest timed out, we'd be left without many of
the "standard" artifacts. Consequently, investigating timed out tests
was particularly unpleasant.

As of this commit, there is parity between the two. We attempt to
collect the same artifacts, but are careful to bound the time we spend
on this (regardless of whether a timeout occurred). Additionally, we are
very careful to leave the cluster in its original state until the
artifacts collection has finished.

Here's the listing of artifacts for `roachtest/hang`, invoked via

```
./pkg/cmd/roachtest/roachstress.sh -c 1 tag:roachtest -- --debug
```

verifying that all artifacts are present. (I also verified that the
`cockroach.log` files contain a goroutine dump).

```
$ find . -maxdepth 3 -not -path './roachprod_state/*' -type f
./teardown.log
./tsdump.gob.yaml
./run_154437.308025163_n1-3_mkdir.log
./test.log
./debug.zip
./__stacks.log
./logs/2.unredacted/cockroach-pebble.td.tobias.2022-02-02T15_41_35Z.148361.log
./logs/2.unredacted/cockroach.stdout.log
./logs/2.unredacted/cockroach.td.tobias.2022-02-02T15_41_35Z.148361.log
./logs/2.unredacted/roachprod.log
./logs/2.unredacted/cockroach.stderr.log
./logs/2.unredacted/cockroach-stderr.td.tobias.2022-02-02T15_41_35Z.148361.log
./logs/2.unredacted/cockroach-health.td.tobias.2022-02-02T15_41_35Z.148361.log
./logs/3.cockroach.log
./logs/3.unredacted/cockroach-stderr.td.tobias.2022-02-02T15_41_35Z.148436.log
./logs/3.unredacted/cockroach.stdout.log
./logs/3.unredacted/cockroach.td.tobias.2022-02-02T15_41_35Z.148436.log
./logs/3.unredacted/roachprod.log
./logs/3.unredacted/cockroach.stderr.log
./logs/3.unredacted/cockroach-pebble.td.tobias.2022-02-02T15_41_35Z.148436.log
./logs/3.unredacted/cockroach-health.td.tobias.2022-02-02T15_41_35Z.148436.log
./logs/2.cockroach.log
./logs/1.cockroach.log
./logs/1.unredacted/cockroach.stdout.log
./logs/1.unredacted/cockroach-pebble.td.tobias.2022-02-02T15_41_34Z.148215.log
./logs/1.unredacted/roachprod.log
./logs/1.unredacted/cockroach-stderr.td.tobias.2022-02-02T15_41_34Z.148215.log
./logs/1.unredacted/cockroach.stderr.log
./logs/1.unredacted/cockroach.td.tobias.2022-02-02T15_41_34Z.148215.log
./logs/1.unredacted/cockroach-health.td.tobias.2022-02-02T15_41_35Z.148215.log
./logs/1.unredacted/cockroach-sql-schema.td.tobias.2022-02-02T15_41_34Z.148215.log
./run_154437.443860723_n1_cockroach_debug_tsdump.log
./run_154437.673769645_n1-3_cockroach_debug_zip.log
./tsdump.gob
./tsdump.gob-run.sh
```

Release note: None

### CODE REVIEW

**Comment by tbg** [comment]
Going to put this on the back burner since https://github.com/cockroachdb/cockroach/pull/73607 shuffles a lot of code around and I'll need to rebase on it.

**Comment by tbg** [comment]
This is now ready for a serious look. I rebased this, polished it up, and made sure it actually works.

File: pkg/cmd/roachtest/test_runner.go [hunk 1]
```diff
@@ -501,8 +501,9 @@ func (r *testRunner) runWorker(
 
 		// Now run the test.
 		l.PrintfCtx(ctx, "starting test: %s:%d", testToRun.spec.Name, testToRun.runNum)
-		var success bool
-		success, err = r.runTest(ctx, t, testToRun.runNum, testToRun.runCount, c, testRunnerLogPath, stdout, testL)
+		err = r.runTest(
```
Comment (otan) [inline pkg/cmd/roachtest/test_runner.go:507]: nit: can we do `if err := ...; err != nil`

**Review by otan — APPROVED** [review]

**Comment by tbg** [comment]
bors r=otan
bors single on

**Pull request merged**